File: immutable.js
function qsort(xs) {
    if (xs.length <= 1) {
        return xs;
    }
    
    const pivot = xs[0];
    const left = xs.slice(1).filter(x => x < pivot);
    const right = xs.slice(1).filter(x => x >= pivot);
    
    return [...qsort(left), pivot, ...qsort(right)];
}

function main() {
    const args = process.argv.slice(2);
    const verbose = args.includes('-v');
    const prePrint = args.includes('-p');
    
    // Generate test data
    const xs = Array.from({length: 10000}, (_, i) => (i * 73 + 19) % 10000);
    
    if (prePrint) {
        console.log('Before sorting:');
        console.log(JSON.stringify(xs));
        return;
    }
    
    const result = qsort(xs);
    
    if (verbose) {
        console.log(JSON.stringify(result));
    } else {
        console.log(JSON.stringify(result.slice(0, 10)));
    }
}

main();

File: immutable.go
package main

import (
	"fmt"
	"os"
)

func qsort(xs []int) []int {
	if len(xs) <= 1 {
		return xs
	}

	pivot := xs[0]
	left := []int{}
	right := []int{}

	for _, x := range xs[1:] {
		if x < pivot {
			left = append(left, x)
		} else {
			right = append(right, x)
		}
	}

	result := qsort(left)
	result = append(result, pivot)
	result = append(result, qsort(right)...)
	return result
}

func main() {
	verbose := false
	prePrint := false
	for _, arg := range os.Args[1:] {
		if arg == "-v" {
			verbose = true
		} else if arg == "-p" {
			prePrint = true
		}
	}
	xs := make([]int, 10000)
	for i := 0; i < 10000; i++ {
		xs[i] = (i*73 + 19) % 10000
	}
	if prePrint {
		fmt.Println("Before sorting:")
		fmt.Println(xs)
		return
	}
	result := qsort(xs)
	if verbose {
		fmt.Println(result)
	} else {
		fmt.Println(result[:10])
	}
}
